Migrate from scss-lint to sass-lint (#7620)

colin-marshall opened the pull request:
This one has been in the works for nearly a year as we waited for sass-lint to mature but it's finally ready thanks to @DanPurdy, @bgriffith, @tonyganch and the @sasstools team! (See #7620 for history)

We drop a ruby dependency with this migration, which might be the last one remaining in Foundation? @zurb/yetinauts please take a look when you get a chance. Not all of the linting rules were the same from scss-lint to sass-lint so feel free to suggest any changes. Check the sass-lint [Rules Documentation](https://github.com/sasstools/sass-lint/tree/master/docs/rules) for further info.

This PR depends on:
- [x] #9335

colin-marshall committed:
Removed gulp-scss-lint and added gulp-sass-lint to package.json

colin-marshall committed:
Replaced scss-lint with sass-lint in gulp lint and sass tasks

colin-marshall committed:
Added .sass-lint.yml config file for sass-lint

DaSchTour commented on gulp/lint.js:
```diff
@@ -13,7 +14,12 @@ gulp.task('lint', ['lint:sass', 'lint:javascript']);
 gulp.task('lint:sass', function() {
   return gulp.src(PATHS)
-    .pipe(scssLint());
+    .pipe(plumber())
```
Just out of curiosity. What is this line doing?

ncoden replied:
`plumber()` make the script ignoring `sass-lint` error and continue the compilation

ncoden replied:
I think it is. It is not "compilation errors", but only "formating warnings". It shouldn't interrupt the compilation.

abdullahsalem commented:
I've not looked at the code yet, but I encourage this orientation.

DanPurdy commented:
Just a mention that the tool to migrate your config looks to be a bit out of date, there's plenty of new rules that should now provide direct equivalents.

Also 1.10 looks to be dropping in the next week or so which adds a few new rules, a whole host of fixes and the ability to 'FINALLY' disable linters via comments.

We're glad it's finally able to happen for you all!

kball commented:
Awesome!  Lets get this to happen for 6.3.  Lets see if we can incorporate @DanPurdy 's feedback and get the config file a little more directly equivalent, but 100% this should be in.

colin-marshall commented:
@DanPurdy you're right the config file is old. I made it back when I first started this branch last year. I will have another go at it. Thanks again for your help.

@kball sounds good! Do you want me to target the 6.3 branch instead then?

kball commented:
@colin-marshall develop is fine - 6.3 will be our next release.  Excited to see you back on github! :)

colin-marshall commented:
@DanPurdy the only rule I could find an equivalent for was `VendorPrefix` -> `no-vendor-prefixes`. Here's what I'm left with:

```yml
# The following scss-lint Linters are not yet supported by sass-lint:
# ElsePlacement, PropertyCount, SelectorDepth, UnnecessaryParentReference
# 
# The following settings/values are unsupported by sass-lint:
# Linter Indentation, option "allow_non_nested_indentation"
# Linter Indentation, option "character"
# Linter PropertySortOrder, option "separate_groups"
# Linter SpaceBeforeBrace, option "allow_single_line_padding"
```

What are some of the other equivalent rules for the above? I see you're still working on PropertyCount. Thanks!

ncoden commented:
@colin-marshall I work on a fork of your branch `colin-marshall:sass-lint` to reorder properties, following https://gist.github.com/ncoden/d42f55df7c7970f548a02cd3468f9c86.

DanPurdy commented:
@colin-marshall There was a bug with the AST that was causing fairly annoying problems with the property count rule in .sass format. It's been on hold for quite a while because of that. The good news is v4.0 of the AST will be landing very very soon which fixes all the remaining issues that blocked it, I'll finish it then. 

Our 1.10 update should be landing within the next week too with even more rules and the long awaited ability to disable lint warnings etc via comments (finally!).

colin-marshall commented:
@ncoden awesome! Updated the PR with your sort order. Just out of curiosity, is this a common/recommended order that other developers use or just what you prefer personally?

@DanPurdy no worries, just wanted to make sure I wasn't missing anything else. Looking forward to 1.10!

ncoden commented on .sass-lint.yml:
```diff
@@ -0,0 +1,347 @@
+    - 
+      order :
+        # Position
+        - position
```
Shouldn't we add `'` (single quotes) on properties names ?

colin-marshall replied:
I went by this example, which uses no quotes.
https://github.com/sasstools/sass-lint/blob/master/docs/rules/property-sort-order.md

Can add them if you think it's necessary.

ncoden replied:
Their internal config files use quotes : https://github.com/sasstools/sass-lint/tree/develop/lib/config/property-sort-orders

I think it is better because of properties are following *external* naming rules (CSS), and not YML ones. Even if in practice it don't change anything.

ncoden commented on .sass-lint.yml:
```diff
@@ -0,0 +1,347 @@
+    - convention: '([a-z0-9]+-?)+'
+  property-sort-order:
+    - 1
+    - ignore-custom-properties: false
```
Please set `ignore-custom-properties` to `true`. We can't handle all CSS properties.

colin-marshall replied:
@ncoden can you elaborate on why this is this please? If this was linting being done in the template I see where setting this to true would be necessary, since the end user can throw any property in there, but this linting will have no effect on what the end user does.

ncoden commented:
@colin-marshall This is my mix between [the tree standards I found](https://github.com/sasstools/sass-lint/tree/develop/lib/config/property-sort-orders) : `smacss` (global organization), `recess` (picked missing properties) and `concentric` (from disposition/ouside to appearance/inside).

So yes, this is sort of "my" sort order. I have too much critics to do to others standards.

ncoden commented:
@colin-marshall Actually, you didn't have to make these changes for `property-sort-order`. This rules wasn't used with `scss-lint` before. You can only migrate to `sass-lint` without too much rules changes, I am adding `property-sort-order` and all the required changes in SCSS files in my branch.

colin-marshall commented:
@ncoden even so I think `property-sort-order` would be useful for us to have here as an easy way to "enforce" consistency in how Foundation's SCSS is written.

ncoden commented:
I'm not sure to understand you. Like I said, I am adding property-sort-order and all the required changes in SCSS files in my branch. So i'm for a consistency too, but changing the lint rules is not the subject of *this* PR.

colin-marshall commented:
@ncoden ohhhh ok. I was confused when you said remove it and thought you meant you were adding the sort orders for your own personal use in your projects and we didn't need to add them here. The fact it came from a gist threw me off. I now see you just meant you were working on the rules in your own branch for a separate PR. My bad! I will revert that commit.

ncoden commented:
@colin-marshall After `position, disposition, margin, padding, global appearance...`, do you prefer:
- `... border, background, shadow ...` (smacss)
- `... shadow, border, background ...` (concentric)
- `... background, border, shadow ...` (recess / classic)

?

(poke @kball)

ncoden commented:
@colin-marshall Don't forget to remove `gem install scss_lint` from `.travis.yml`
https://github.com/zurb/foundation-sites/blob/6c7b0b4af5502ca5e2495f9c50caedb398bd2d12/.travis.yml#L9

colin-marshall committed:
Added no-vendor-prefixes linter to sass-lint.yml

In scss-lint the VendorPrefix linter was incorrectly titled VendorPrefixes so the rules did not migrate properly to sass-lint.

colin-marshall committed:
Remove scss-lint.yml and the scss-lint install script

No longer necessary after migrating to sass-lint.

colin-marshall commented on .travis.yml:
```diff
@@ -1,9 +1,8 @@
-  - "0.12"
+  - "4.0"
 before_install:
   - rvm install 2.2.2
```
@zurb/yetinauts does anybody know if we need this `rvm install` script for anything besides scss-lint? If that's all it's being used for I think I will just remove it in this PR.

ncoden replied:
There is no other possible uses of rvm than for our own gems, isn't it ?

colin-marshall replied:
@ncoden none that I'm aware of but I am not 100% sure that scss-lint is the last thing Ruby related in the repo so I wanted to verify that before removing rvm in the travis scripts.

ncoden commented on .sass-lint.yml:
```diff
@@ -0,0 +1,128 @@
+    - 1
+    - allow-leading-underscore: true
+      convention: hyphenatedlowercase
+  zero-unit: 1
```
Line end

colin-marshall replied:
Fixed. Thanks!

colin-marshall committed:
Remove rvm install script from travis.yml

No longer necessary since we dropped our last ruby dependency when we migrated to sass-lint from scss-lint

colin-marshall committed:
Added blank space between sass-lint rules

colin-marshall committed:
Removed Ruby references in README

ncoden commented:
:tada:
